cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -856,11 +856,15 @@ private predicate unary_simple_comparison_eq(
     inNonZeroCase = false
   )
   or
-  // There's no implicit CompareInstruction in files compiled as C since C
-  // doesn't have implicit boolean conversions. So instead we check whether
-  // there's a branch on a value of pointer or integer type.
-  relevantUnaryComparison(test) and
-  op.getDef() = test and
+  not test.isGLValue() and
+  not simple_comparison_eq(test, _, _, _, _) and
+  not simple_comparison_lt(test, _, _, _) and
```

Comment:
The exclusion of switch expressions I understand, as these are handled above, but why are these three lines the correct thing to do?

Reply:
It doesn't really matter. Without these conjuncts, this predicate basically consists of all instructions of type address, integer, or boolean. This is simply a very easy way of reducing this relatively large set.

Reply:
Thanks.

> by excluding instructions which are already guard conditions for other reasons (i.e, by being included in `simple_comparison_eq` or `simple_comparison_lt `), or being irrelevant (i.e., by being a glvalue).

This is indeed what I expect. However, I have a hard time following the logic of the library here. For example, I started looking at the places where `unary_simple_comparison_eq` is used, and expected `simple_comparison_eq` or `simple_comparison_lt` or `isGLValue` there, but didn't see those predicates mentioned.

Reply:
Yeah, I can see why this is hard to follow. `simple_comparison_eq` is used in another code path (when we infer `e1 == e2 + k` equalities) in [compares_eq](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L716) (which is used when a query uses [GuardCondition::comparesEq](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L558)), and `simple_comparison_lt` is used in _yet another_ code path (where we infer `e1 < e2 + k` inequalities) when a query uses [GuardCondition::comparesLt](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L492).

I'll write up a comment explaining this in this case. And now that I'm here I'll write some more general comments in this file as well.

Reply:
I've added some explanatory comments to 7819cc1c36fda57fdac1421f9cf8b32c570f366e. Does that help in this case?

Reply:
Thanks. That helps a lot.